flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/graph/TestJobBuilders.java

```diff
@@ -0,0 +1,292 @@
+                                                    unitedSourceLeft, eventQueue, commandQueue))
+                                    .setUidHash(unitedSourceLeft)
+                                    .assignTimestampsAndWatermarks(createWmAssigner());
+
+                    SingleOutputStreamOperator<TestDataElement> forwardTransform =
+                            src.transform(
+                                            "transform-1-forward",
+                                            TypeInformation.of(TestDataElement.class),
+                                            new OneInputTestStreamOperatorFactory(
+                                                    mapForward, eventQueue))
+                                    .setUidHash(mapForward);
+
+                    forwardTransform.addSink(new DiscardingSink<>());
+
+                    Map<String, Integer> operatorsNumberOfInputs = new HashMap<>();
+                    operatorsNumberOfInputs.put(mapForward, 1);
+
+                    return new TestJobWithDescription(
+                            env.getStreamGraph().getJobGraph(),
+                            singleton(unitedSourceLeft),
+                            new HashSet<>(singletonList(mapForward)),
+                            new HashSet<>(asList(unitedSourceLeft, mapForward)),
+                            operatorsNumberOfInputs,
+                            eventQueue,
+                            commandQueue);
+                }
+
+                @Override
+                public String toString() {
+                    return "simple graph";
+                }
+            };
+
+    public static final TestingGraphBuilder COMPLEX_GRAPH_BUILDER =
+            new TestingGraphBuilder() {
+                @Override
+                public TestJobWithDescription apply(
+                        SharedObjects shared,
+                        Consumer<Configuration> confConsumer,
+                        Consumer<StreamExecutionEnvironment> envConsumer) {
+
+                    TestEventQueue eventQueue = TestEventQueue.createShared(shared);
+                    TestCommandDispatcher commandQueue = TestCommandDispatcher.createShared(shared);
+
+                    StreamExecutionEnvironment env = prepareEnv(confConsumer, envConsumer);
+
+                    // using hashes so that operators emit identifiable events
+                    String unitedSourceLeft = OP_ID_HASH_PREFIX + "1";
+                    String unitedSourceRight = OP_ID_HASH_PREFIX + "2";
+                    String connectedSource = OP_ID_HASH_PREFIX + "3";
+                    String multiSource = OP_ID_HASH_PREFIX + "4";
+                    String mapForward = OP_ID_HASH_PREFIX + "5";
+                    String mapKeyed = OP_ID_HASH_PREFIX + "6";
+                    String mapTwoInput = OP_ID_HASH_PREFIX + "7";
+                    String multipleInput = OP_ID_HASH_PREFIX + "8";
+
+                    // todo: FLIP-27 sources
```

Comment:
What is the plan for this?

Reply:
If there is any left over TODO, let's do it after merging @rkhachatryan's PR and let's plan it between me, you @dawidwys  and @gaoyunhaii .